pygac/gac_reader.py

```diff
@@ -232,3 +241,235 @@ def get_angles(self):
+
+        LOG.debug('Removed {0} scanline(s) with corrupt scanline numbers'
+                  .format(len(along_track) - len(self.scans)))
+
+        # Plot corrected scanline numbers
+        if plot:
+            along_track = along_track[within_range]
+            along_track = along_track[diffs <= thresh]
+            ax0.plot(along_track, self.scans["scan_line_number"], "r--",
+                     label="corrected")
+            ax0.set_ylabel("Scanline Number")
+            ax0.set_xlabel("Along Track")
+            ax0.legend(loc="best")
+
+            ax1.plot(np.arange(len(nz_diffs)), nz_diffs)
+            ax1.axhline(thresh, color="r", label="thresh={0:.2f}"
+                        .format(thresh))
+            ax1.set_xlabel("Index")
+            ax1.set_ylabel("nonzero |n - n'|")
+            ax1.legend()
+
+            plt.tight_layout()
+            plt.savefig(self.filename + "_scanline_number_correction.png",
+                        bbox_inches='tight')
+
+    def correct_utcs(self, max_diff_from_t0_head=6*60*1000,
+                     min_frac_near_t0_head=0.01, max_diff_from_ideal_t=10*1000,
+                     plot=False):
+        """
+        Correct corrupt scanline timestamps using the scanline number and an
+        approximate scanning rate of 2 lines per second.
+
+        The header timestamp is used as a guideline to estimate the offset
+        between timestamps computed from the scanline number and the actual
+        scanline timestamps in the data. If header timestamp and scanline
+        timestamps do not match, no correction is applied.
+
+        Once the offset has been estimated, one can calculate the ideal
+        timestamps based on the scanline number. Timestamps deviating more than
+        a certain threshold from the ideal timestamps are replaced by
+        the ideal timestamps.
+
+        Example files having corrupt timestamps:
+            - NSS.GHRR.NA.D81193.S2329.E0116.B1061214.WI
+            - NSS.GHRR.NL.D01035.S2342.E0135.B0192627.WI
+
+        @param max_diff_from_t0_head: Threshold for offset estimation: A
+        scanline timestamp matches the header timestamp t0_head if it is
+        within the interval
+
+          [t0_head - max_diff_from_t0_head, t0_head + max_diff_from_t0_head]
+
+        around the header timestamp.
+        @param min_frac_near_t0_head: Specifies the minimum fraction of
+        scanline timestamps matching the header timestamp required for
+        applying the correction.
+        @param max_diff_from_ideal_t: Threshold for timestamp correction: If
+        a scanline timestamp deviates more than max_diff_from_ideal_t from
+        the ideal timestamp, it is regarded as corrupt and will be replaced with
+        the ideal timestamp.
+        @param plot: If True, plot results.
+        """
+        apply_corr = True
+        fail_reason = ""
+        scanning_rate = 2.0 / 1000.0  # scanlines per millisecond
```

Comment:
Ok, I will fix those myself when I merge lac and gac then. We can discuss later the relevance of the correction for lac :)